File: src/sum_multi.go
// sum_multi.go - Go multi-threaded sum
package main

import (
    "fmt"
    "runtime"
    "sync"
    "time"
)

func main() {
    start := time.Now()
    
    const n = 100_000_000
    numWorkers := runtime.NumCPU()
    chunkSize := n / numWorkers
    
    var wg sync.WaitGroup
    results := make([]int64, numWorkers)
    
    fmt.Printf("Using %d goroutines\n", numWorkers)
    
    for i := 0; i < numWorkers; i++ {
        wg.Add(1)
        go func(workerID int) {
            defer wg.Done()
            
            start := int64(workerID*chunkSize + 1)
            end := int64((workerID+1) * chunkSize)
            if workerID == numWorkers-1 {
                end = n // Handle remainder
            }
            
            var localSum int64 = 0
            for j := start; j <= end; j++ {
                localSum += j
            }
            results[workerID] = localSum
        }(i)
    }
    
    wg.Wait()
    
    var totalSum int64 = 0
    for _, result := range results {
        totalSum += result
    }
    
    elapsed := time.Since(start)
    fmt.Printf("Sum: %d\n", totalSum)
    fmt.Printf("Go Multi-threaded Time: %v\n", elapsed)
    
    // Verify correctness
    const expected int64 = 5000000050000000
    if totalSum == expected {
        fmt.Println("✅ Result is correct!")
    } else {
        fmt.Printf("❌ Expected %d, got %d\n", expected, totalSum)
    }
}

File: src/sum_multi.rs
use std::sync::mpsc;
use std::thread;
use std::time::Instant;

fn main() {
    let start = Instant::now();

    let n: i64 = 100_000_000;
    let num_threads = num_cpus::get();
    let chunk_size = n / num_threads as i64;

    println!("Using {} threads", num_threads);

    let (tx, rx) = mpsc::channel();

    for i in 0..num_threads {
        let tx_clone = tx.clone();
        thread::spawn(move || {
            let start_val = (i as i64) * chunk_size + 1;
            let end_val = if i == num_threads - 1 {
                n
            } else {
                (i as i64 + 1) * chunk_size
            };

            let mut local_sum: i64 = 0;
            for j in start_val..=end_val {
                local_sum += j;
            }

            tx_clone.send(local_sum).unwrap();
        })
    }
}

drop(tx);

let total_sum: i64 = rx.iter().sum();

let elapsed = start.elapsed();
println!("Sum: {}", total_sum);
println!("Rust Multi-threaded Time: {:?}", elapsed);

// Verify correctness
let expected: i64 = 5000000050000000;
if total_sum == expected {
    println!("✅ Result is correct!");
} else {
    println!("❌ Expected {}, got {}", expected, total_sum);
    std::process::exit(1);
}